skimage/morphology/misc.py

```diff
@@ -235,3 +243,212 @@ def remove_small_holes(ar, area_threshold=64, connectivity=1, *, out=None):
+    ... )
+    >>> ski.morphology.remove_objects_by_distance(
+    ...     label_image, min_distance=3
+    ... )
+    array([[8, 0, 0, 0, 0, 0, 0, 0, 0, 9, 9],
+           [8, 8, 8, 0, 0, 0, 0, 0, 0, 9, 9],
+           [0, 0, 0, 0, 0, 0, 0, 0, 9, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+           [2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 7]])
+    """
+    if min_distance < 0:
+        raise ValueError(f"min_distance must be >= 0, was {min_distance}")
+    if not np.issubdtype(label_image.dtype, np.integer):
+        raise ValueError(
+            f"`label_image` must be of integer dtype, got {label_image.dtype}"
+        )
+    if out is None:
+        out = label_image.copy(order="C")
+    elif out is not label_image:
+        out[:] = label_image
+    # May create a copy if order is not C, account for that later
+    out_raveled = out.ravel(order="C")
+
+    if spacing is not None:
+        spacing = np.array(spacing)
+        if spacing.shape != (out.ndim,) or spacing.min() <= 0:
+            raise ValueError(
+                "`spacing` must contain exactly one positive factor "
+                "for each dimension of `label_image`"
+            )
+
+    indices = np.nonzero(out_raveled)[0]
+    # Optimization: Split indices into those on the object boundaries and inner
+    # ones. The KDTree is built only from the boundary indices, which reduces
+    # the size of the critical loop significantly! Remaining indices are only
+    # used to remove the inner parts of objects as well.
+    if (spacing is None or np.all(spacing[0] == spacing)) and p_norm <= 2:
+        # For unity spacing we can make the borders more sparse by using a
+        # lower connectivity
+        footprint = ndi.generate_binary_structure(out.ndim, 1)
+    else:
+        footprint = ndi.generate_binary_structure(out.ndim, out.ndim)
+    border = (
+        ndi.maximum_filter(out, footprint=footprint)
+        != ndi.minimum_filter(out, footprint=footprint)
+    ).ravel()[indices]
+    border_indices = indices[border]
+    inner_indices = indices[~border]
+
+    if border_indices.size == 0:
+        # Image without any or only one object, return early
+        return out
+
+    # Sort by label ID first, so that IDs of the same object are contiguous
+    # in the sorted index. This allows fast discovery of the whole object by
+    # simple iteration up or down the index!
+    border_indices = border_indices[np.argsort(out_raveled[border_indices])]
+    inner_indices = inner_indices[np.argsort(out_raveled[inner_indices])]
+
+    if priority is None:
+        if out_raveled.dtype.itemsize > np.intp().itemsize:
+            # bincount expects intp (e.g. 32-bit on WASM), so down-cast to that
+            priority = np.bincount(out_raveled.astype(np.intp, copy=False))
+        else:
+            priority = np.bincount(out_raveled)
```

Comment:
Okay, I tried to address the WASM-bincount failure in https://github.com/scikit-image/scikit-image/pull/4165/commits/e6446d6fd2b2674ab361de344b2172323f673fd1. Not sure if this is the best approach though. I wanted to avoid creating unnecessary up-casting of int8 etc.

Reply:
Thanks for chiming in @agriyakhetarpal & @seberg.

Reply:
That didn't work. https://github.com/scikit-image/scikit-image/pull/4165/commits/d749dd316d40fd7d1ca71c74d170fd16b2c5143d works though.

Reply:
Out of curiosity, what if the values are too large to be casts down to int32?

Reply:
I'm confused — aren't there non-wasm platforms where this would be an issue? Or is it because we no longer support 32-bit win or linux that we can get away with just is_wasm?

Reply:
Yeah, this just force casts which is unfortunately something that indexing-like operations do in NumPy, so if you manage to end up with a too large value, that could silently do the wrong (although I guess it's likely that even after the force cast you have huge values that cause an out of memory error).

If floats might come around using `casting="same_kind"` would be better; you can do this unconditionally really, that is what the `copy=False` is for.